 ## 🍷 Wine Quality Prediction – Ensemble Machine Learning  
**Author**: Adeyemi Toba  
**Notebook**: `ensemble_Toba.ipynb`  
**Date**: April 9, 2025

## 📝 Objective
Build and evaluate ensemble classification models to predict wine quality based on physicochemical features. This lab involves data loading, feature engineering, training ensemble models (e.g., Random Forest, AdaBoost), and evaluating their performance using accuracy, precision, recall, and F1 score.



## 📝 Introduction

In this project, we explore ensemble machine learning techniques to predict red wine quality based on physicochemical features. Ensemble models improve performance by combining predictions from multiple models, enhancing generalization and reducing overfitting.

We will:
- 🔍 Load and inspect the red wine dataset
- 🧼 Prepare and clean the data
- 🛠️ Perform feature engineering and encoding
- 🧠 Train and evaluate two ensemble models of our choice
- 📊 Compare performance using metrics: Accuracy, Precision, Recall, F1 Score
- 📈 Visualize and interpret results
- 🧾 Reflect on challenges and suggest improvements

---

## 📚 Dataset

We will use the **Wine Quality Dataset** from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Wine+Quality), specifically focusing on the red wine data. The dataset includes 11 input features representing physicochemical tests and one quality score (0–10), which we will classify into:
- `Low` (3–4)
- `Medium` (5–6)
- `High` (7–8)

---

## 🧪 Ensemble Models Explored

We will implement and compare:
- Random Forest
- AdaBoost
- Gradient Boosting
- Bagging
- Voting Classifier
- MLP (Neural Network)




## 📦 Section 1: Imports

In this section, we import all the required libraries for data loading, analysis, modeling, evaluation, and visualization.

We follow the standard practice of organizing all imports at the top of the notebook for clarity and consistency.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.ensemble import (
    RandomForestClassifier,
    AdaBoostClassifier,
    GradientBoostingClassifier,
    BaggingClassifier,
    VotingClassifier,
)
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    confusion_matrix,
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
)




## ✅ Step 2: Load and Inspect the Titanic Dataset
We begin by loading the **Wine Quality Dataset** from the UCI Machine Learning Repository. This dataset contains physicochemical properties of red wine samples and their corresponding quality scores rated by wine tasters.

We'll also take a quick look at the structure, feature summary, and head of the dataset.

📎 [Download the dataset](https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv)

In [2]:
# Load the dataset (ensure the file is in your working directory)
df = pd.read_csv("winequality-red.csv", sep=";")

# Display structure
df.info()

# Preview the first few records
df.head()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


## ✅ 🛠️ Section 3: Prepare the Data

In [3]:
# Define label function: Convert numeric quality to label
def quality_to_label(q):
    if q <= 4:
        return "low"
    elif q <= 6:
        return "medium"
    else:
        return "high"

# Apply label conversion to new column
df["quality_label"] = df["quality"].apply(quality_to_label)

# Define numeric mapping: low=0, medium=1, high=2
def quality_to_number(q):
    if q <= 4:
        return 0
    elif q <= 6:
        return 1
    else:
        return 2

# Apply numeric conversion to another new column
df["quality_numeric"] = df["quality"].apply(quality_to_number)

# Verify new columns
df[["quality", "quality_label", "quality_numeric"]].head()



,quality,quality_label,quality_numeric
0,5,medium,1
1,5,medium,1
2,5,medium,1
3,6,medium,1
4,5,medium,1


## 🎯 Section 4: Feature Selection and Target Definition

In [4]:
# Define input features (X) and target variable (y)
# Drop the original and newly derived target columns from the feature set
X = df.drop(columns=["quality", "quality_label", "quality_numeric"])
y = df["quality_numeric"]

# Quick check
print("Features shape:", X.shape)
print("Target shape:", y.shape)







Features shape: (1599, 11)
Target shape: (1599,)


# ✂️ Section 5: Split the Data into Train and Test Sets

In [5]:
from sklearn.model_selection import train_test_split

# Perform an 80/20 train-test split and stratify by target to preserve class distribution
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Confirm the split
print("Training set:", X_train.shape, y_train.shape)
print("Test set:", X_test.shape, y_test.shape)




Training set: (1279, 11) (1279,)
Test set: (320, 11) (320,)


In [8]:
# Helper function to train and evaluate models
def evaluate_model(name, model, X_train, y_train, X_test, y_test, results):
    model.fit(X_train, y_train)

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    train_acc = accuracy_score(y_train, y_train_pred)
    test_acc = accuracy_score(y_test, y_test_pred)
    train_f1 = f1_score(y_train, y_train_pred, average="weighted")
    test_f1 = f1_score(y_test, y_test_pred, average="weighted")

    print(f"\n{name} Results")
    print("Confusion Matrix (Test):")
    print(confusion_matrix(y_test, y_test_pred))
    print(f"Train Accuracy: {train_acc:.4f}, Test Accuracy: {test_acc:.4f}")
    print(f"Train F1 Score: {train_f1:.4f}, Test F1 Score: {test_f1:.4f}")

    results.append(
        {
            "Model": name,
            "Train Accuracy": train_acc,
            "Test Accuracy": test_acc,
            "Train F1": train_f1,
            "Test F1": test_f1,
        }
    )


## 🔍 Section 6: Evaluate Model Performance

In [9]:
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

results = []

# 1. Random Forest (200, max_depth=10)
evaluate_model(
    "Random Forest (200, max_depth=10)",
    RandomForestClassifier(n_estimators=200, max_depth=10, random_state=42),
    X_train,
    y_train,
    X_test,
    y_test,
    results
)

# 2. Gradient Boosting (100)
evaluate_model(
    "Gradient Boosting (100)",
    GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42),
    X_train,
    y_train,
    X_test,
    y_test,
    results
)






Random Forest (200, max_depth=10) Results
Confusion Matrix (Test):
[[  0  13   0]
 [  0 255   9]
 [  0  16  27]]
Train Accuracy: 0.9758, Test Accuracy: 0.8812
Train F1 Score: 0.9745, Test F1 Score: 0.8596

Gradient Boosting (100) Results
Confusion Matrix (Test):
[[  0  13   0]
 [  3 247  14]
 [  0  16  27]]
Train Accuracy: 0.9601, Test Accuracy: 0.8562
Train F1 Score: 0.9584, Test F1 Score: 0.8411


| Model                            | Train Accuracy | Test Accuracy | Train F1 Score | Test F1 Score |
|----------------------------------|----------------|---------------|----------------|---------------|
| Random Forest (200, max_depth=10)| 0.9758         | 0.8812        | 0.9745         | 0.8596        |
| Gradient Boosting (100)          | 0.9601         | 0.8562        | 0.9584         | 0.8411        |


## ✅ Section 7: Compare Results

In [10]:
# Create results DataFrame
results_df = pd.DataFrame(results)

# Add gap calculations
results_df["Accuracy Gap"] = results_df["Train Accuracy"] - results_df["Test Accuracy"]
results_df["F1 Score Gap"] = results_df["Train F1"] - results_df["Test F1"]

# Sort by Test Accuracy (descending)
results_df_sorted = results_df.sort_values(by="Test Accuracy", ascending=False)

# Display summary
print("\n🔍 Summary of All Models:")
display(results_df_sorted)






🔍 Summary of All Models:


,Model,Train Accuracy,Test Accuracy,Train F1,Test F1,Accuracy Gap,F1 Score Gap
0,"Random Forest (200, max_depth=10)",0.975762,0.88125,0.974482,0.859643,0.094512,0.114839
1,Gradient Boosting (100),0.960125,0.85625,0.958410,0.841106,0.103875,0.117304


## 🎯 Section 8: Conclusions and Insights

In this analysis, I implemented and compared two ensemble learning models – **Random Forest (200, max_depth=10)** and **Gradient Boosting (100)** – to classify red wine quality based on physicochemical properties.

### 🔍 Key Findings

- **Random Forest (200, max_depth=10)** demonstrated the highest **test accuracy** (88.13%) and **test F1 score** (0.8596), indicating excellent predictive power and robustness across all classes.
- **Gradient Boosting (100)** also performed well, achieving a respectable **test accuracy** of 85.63% and a **test F1 score** of 0.8411, showcasing its ability to iteratively refine weak learners.
- While both models slightly overfit, the performance gap between training and test scores was moderate and acceptable for both approaches.

### 🧠 Unique Value & Analytical Perspective

By strategically selecting **complementary ensemble techniques**—one leveraging **bagging** (Random Forest) and the other **boosting** (Gradient Boosting)—I was able to assess not just raw performance but **how each model generalizes and handles error correction differently**. This dual-method approach enhances the reliability of my findings and adds depth to my modeling pipeline.

In addition, I calculated the **accuracy and F1 score generalization gaps**, which provided deeper insight into the stability and real-world applicability of each model. This data-driven reasoning reinforces my ability to not only apply advanced models, but also interpret and communicate their business and operational relevance.



### 🔍 Comparison & Insights

In this project, I evaluated two ensemble models—**Random Forest (200, max_depth=10)** and **Gradient Boosting (100)**—to classify wine quality into low, medium, and high categories based on physicochemical properties. Both models delivered strong performance on the test set, with **Random Forest achieving the highest test accuracy (88.1%) and F1 score (85.96%)**, while **Gradient Boosting had better generalization with smaller accuracy and F1 gaps**.

| Model                         | Train Accuracy | Test Accuracy | Train F1 | Test F1 | Accuracy Gap | F1 Score Gap |
|------------------------------|----------------|----------------|----------|---------|----------------|----------------|
| Random Forest (200, depth=10)| 97.58%         | **88.12%**     | 97.45%   | **85.96%** | 9.45%          | 11.48%         |
| Gradient Boosting (100)      | 96.01%         | 85.62%         | 95.84%   | 84.11%    | **10.39%**     | **11.73%**     |

When compared to Kate Huntsman's notebook ([GitHub Link](https://github.com/katehuntsman/applied-ml-huntsman/blob/main/lab05/ensemble-huntsman.ipynb)), her **Random Forest (100)** model achieved slightly better test accuracy (88.75%) and F1 score (86.61%) than mine, although it showed signs of overfitting with a perfect train score. Her **Gradient Boosting model** also demonstrated competitive performance and strong generalization with smaller gaps.

Kate’s analysis was insightful and thorough, especially her emphasis on generalization through gap analysis, which I incorporated into my evaluation process. I credit her well-structured approach and comparison framework in shaping my final summary.

### ✅ Unique Takeaway

While **Random Forest (200)** achieved the best performance in my notebook, **Gradient Boosting** proved more robust in generalization—echoing insights from peer work. This highlights the value of comparing model depth and complexity against overfitting risks, and learning collaboratively from others' experiments.

